File: agents/src/agents/architect_board.py
"""Runtime state of what the Architect has drawn, plus the pixel math it uses
to place things. The model works with columns, ids and kinds; every coordinate
is worked out here so it stays consistent between ops and revisions."""

from dataclasses import dataclass, field
from typing import Optional

from agents.state import NodeKind

# Layout constants. Same numbers canvas.py has used since day one; the sticky
# size mirrors NOTE_SIZE in boardElements.ts, so an arrow always lands on the
# sticky the front actually drew.
ORIGIN_X = 210
ORIGIN_Y = 104
COLUMN_W = 200
COLUMN_GAP = 64
NOTE_W = 160
NOTE_H = 110
NOTE_GAP = 48
GROUP_HEAD = 52
GROUP_PAD = 20
ANNOTATION_H = 68
ANNOTATION_STEP = ANNOTATION_H + 8
ANNOTATION_GAP = 12
INK = "#525252"

# Kind → sticky style. Same palette canvas.py uses today, kept here so this
# module can produce a full element snapshot without importing canvas.
TONE: dict[str, tuple[str, str, str]] = {
    "idea": ("Idea", "amber", "#f59e0b"),
    "decision": ("Decisión", "green", "#22c55e"),
    "tarea": ("Tarea", "blue", "#3b82f6"),
    "duda": ("Duda", "violet", "#8b5cf6"),
}


@dataclass
class Node:
    id: str
    texto: str
    columna: int
    row: int
    kind: NodeKind


@dataclass
class Annotation:
    id: str
    nodo_id: str
    texto: str
    autor: str
    slot: int


@dataclass
class Arrow:
    id: str
    de: str
    a: str
    label: str


@dataclass
class ArchitectBoard:
    """Everything the Architect currently has on the pizarra, keyed for O(1)
    lookup by id. The private stacks preserve visual order per column and per
    node so a new sticky lands under the last one instead of on top of it."""

    nodes: dict[str, Node] = field(default_factory=dict)
    annotations: dict[str, Annotation] = field(default_factory=dict)
    arrows: dict[str, Arrow] = field(default_factory=dict)
    titles: dict[int, str] = field(default_factory=dict)
    _column_stacks: dict[int, list[str]] = field(default_factory=dict)
    _note_stacks: dict[str, list[str]] = field(default_factory=dict)


def column_x(columna: int) -> float:
    return ORIGIN_X + columna * (COLUMN_W + COLUMN_GAP) + GROUP_PAD


def _column_head(board: ArchitectBoard, columna: int) -> float:
    # Only columns with a title get the tall header. An untitled column starts
    # right at ORIGIN_Y — the alternative is a floating gap that reads as a bug.
    return GROUP_HEAD if board.titles.get(columna) else 0


def node_xy(board: ArchitectBoard, columna: int, row: int) -> tuple[float, float]:
    return (
        column_x(columna),
        ORIGIN_Y + _column_head(board, columna) + row * (NOTE_H + NOTE_GAP),
    )


def annotation_xy(node_x: float, node_y: float, slot: int) -> tuple[float, float]:
    return (node_x + NOTE_W + ANNOTATION_GAP, node_y + slot * ANNOTATION_STEP)


def title_xy(columna: int) -> tuple[float, float]:
    return (column_x(columna), ORIGIN_Y + 12)


def _arrow_points(
    board: ArchitectBoard, source: Node, target: Node
) -> tuple[float, float, float, float]:
    """Where a line between two nodes actually starts and ends. Down the column
    when both live in it, across when they do not — an arrow that cuts through
    the notes in between reads as a mistake."""
    sx, sy = node_xy(board, source.columna, source.row)
    tx, ty = node_xy(board, target.columna, target.row)
    if source.columna == target.columna:
        return (sx + NOTE_W / 2, sy + NOTE_H, tx + NOTE_W / 2, ty)
    side = 0 if sx < tx else NOTE_W
    return (sx + (NOTE_W - side), sy + NOTE_H / 2, tx + side, ty + NOTE_H / 2)


# --- op application. Each function returns the wire payload or None if the op
# was a no-op (target missing, duplicate id, unknown kind). Returning None is
# the way the Architect's mistakes are swallowed without breaking the board.


def crear_nodo(
    board: ArchitectBoard, id_: str, texto: str, columna: int, kind: str
) -> Optional[dict]:
    if not id_ or not texto or kind not in TONE:
        return None
    if id_ in board.nodes:
        return None
    stack = board._column_stacks.setdefault(columna, [])
    row = len(stack)
    stack.append(id_)
    board.nodes[id_] = Node(id_, texto, columna, row, kind)  # type: ignore[arg-type]
    x, y = node_xy(board, columna, row)
    return {
        "type": "crear_nodo",
        "id": id_,
        "texto": texto,
        "columna": columna,
        "kind": kind,
        "x": x,
        "y": y,
    }


def editar_nodo(board: ArchitectBoard, id_: str, texto: str) -> Optional[dict]:
    node = board.nodes.get(id_)
    if not node or not texto or texto == node.texto:
        return None
    node.texto = texto
    return {"type": "editar_nodo", "id": id_, "texto": texto}


def mover_nodo(board: ArchitectBoard, id_: str, columna: int) -> Optional[dict]:
    node = board.nodes.get(id_)
    if not node or node.columna == columna:
        return None
    old = board._column_stacks.get(node.columna, [])
    if id_ in old:
        old.remove(id_)
    new_stack = board._column_stacks.setdefault(columna, [])
    node.columna = columna
    node.row = len(new_stack)
    new_stack.append(id_)
    x, y = node_xy(board, columna, node.row)
    return {"type": "mover_nodo", "id": id_, "columna": columna, "x": x, "y": y}


def conectar(
    board: ArchitectBoard, de: str, a: str, label: str
) -> Optional[dict]:
    if de == a or de not in board.nodes or a not in board.nodes:
        return None
    arrow_id = f"a_{de}_{a}"
    if arrow_id in board.arrows:
        return None
    board.arrows[arrow_id] = Arrow(arrow_id, de, a, label or "")
    x1, y1, x2, y2 = _arrow_points(board, board.nodes[de], board.nodes[a])
    op: dict = {
        "type": "conectar",
        "id": arrow_id,
        "de": de,
        "a": a,
        "x1": x1,
        "y1": y1,
        "x2": x2,
        "y2": y2,
    }
    if label:
        op["label"] = label
    return op


def pegar_nota(
    board: ArchitectBoard, nodo_id: str, texto: str, autor: str
) -> Optional[dict]:
    if nodo_id not in board.nodes or not texto:
        return None
    stack = board._note_stacks.setdefault(nodo_id, [])
    slot = len(stack)
    ann_id = f"n_{nodo_id}_{slot}"
    stack.append(ann_id)
    board.annotations[ann_id] = Annotation(
        ann_id, nodo_id, texto, autor or "", slot
    )
    node = board.nodes[nodo_id]
    nx, ny = node_xy(board, node.columna, node.row)
    x, y = annotation_xy(nx, ny, slot)
    op: dict = {
        "type": "pegar_nota",
        "id": ann_id,
        "nodo_id": nodo_id,
        "texto": texto,
        "x": x,
        "y": y,
    }
    if autor:
        op["autor"] = autor
    return op


def borrar(board: ArchitectBoard, id_: str) -> Optional[dict]:
    if id_ in board.nodes:
        node = board.nodes.pop(id_)
        stack = board._column_stacks.get(node.columna, [])
        if id_ in stack:
            stack.remove(id_)
        # A node's annotations and arrows lose their anchor when the node goes,
        # so they leave with it — otherwise the board fills up with things
        # hanging off nothing.
        for ann_id in list(board._note_stacks.pop(id_, [])):
            board.annotations.pop(ann_id, None)
        orphans = [
            aid for aid, ar in board.arrows.items() if ar.de == id_ or ar.a == id_
        ]
        for aid in orphans:
            board.arrows.pop(aid, None)
        return {"type": "borrar", "id": id_}

    if id_ in board.annotations:
        ann = board.annotations.pop(id_)
        stack = board._note_stacks.get(ann.nodo_id, [])
        if id_ in stack:
            stack.remove(id_)
        return {"type": "borrar", "id": id_}

    if id_ in board.arrows:
        board.arrows.pop(id_)
        return {"type": "borrar", "id": id_}

    return None


def titular_columna(
    board: ArchitectBoard, columna: int, titulo: str
) -> Optional[dict]:
    if not titulo or board.titles.get(columna) == titulo:
        return None
    board.titles[columna] = titulo
    x, y = title_xy(columna)
    return {
        "type": "titular_columna",
        "columna": columna,
        "titulo": titulo,
        "x": x,
        "y": y,
    }


def snapshot(board: ArchitectBoard) -> dict:
    """The structured board as JSON, for a browser opening the session late.
    Every field the frontend reducer would have built up from ops, minus the
    private stacks — those are backend bookkeeping."""
    return {
        "nodes": [
            {
                "id": n.id,
                "texto": n.texto,
                "columna": n.columna,
                "kind": n.kind,
                **dict(zip(("x", "y"), node_xy(board, n.columna, n.row))),
            }
            for n in board.nodes.values()
        ],
        "annotations": [
            {
                "id": a.id,
                "nodo_id": a.nodo_id,
                "texto": a.texto,
                "autor": a.autor,
                **dict(
                    zip(
                        ("x", "y"),
                        annotation_xy(
                            *node_xy(
                                board,
                                board.nodes[a.nodo_id].columna,
                                board.nodes[a.nodo_id].row,
                            ),
                            a.slot,
                        ),
                    )
                ),
            }
            for a in board.annotations.values()
            if a.nodo_id in board.nodes
        ],
        "arrows": [
            {
                "id": ar.id,
                "de": ar.de,
                "a": ar.a,
                "label": ar.label,
                **dict(
                    zip(
                        ("x1", "y1", "x2", "y2"),
                        _arrow_points(board, board.nodes[ar.de], board.nodes[ar.a]),
                    )
                ),
            }
            for ar in board.arrows.values()
            if ar.de in board.nodes and ar.a in board.nodes
        ],
        "titles": [
            {
                "columna": columna,
                "titulo": titulo,
                **dict(zip(("x", "y"), title_xy(columna))),
            }
            for columna, titulo in board.titles.items()
            if titulo
        ],
    }


def to_elements(board: ArchitectBoard) -> list[dict]:
    """The whole board as canvas primitives, for consumers that expect the flat
    list (a browser opening a session that is already halfway through)."""
    elements: list[dict] = []

    def base(id_: str, color: str, **rest) -> dict:
        return {
            "id": id_,
            "color": color,
            "width": 2,
            "opacity": 100,
            "dash": "solid",
            "radius": 8,
            "cap": "round",
            **rest,
        }

    for columna, titulo in board.titles.items():
        if not titulo:
            continue
        x, y = title_xy(columna)
        elements.append(
            base(f"ft{columna}", INK, kind="text", x=x, y=y, text=titulo)
        )

    for node in board.nodes.values():
        x, y = node_xy(board, node.columna, node.row)
        tag, tone, colour = TONE[node.kind]
        elements.append(
            base(
                f"n{node.id}",
                colour,
                kind="note",
                x=x,
                y=y,
                text=node.texto,
                tag=tag,
                tone=tone,
            )
        )

    for ann in board.annotations.values():
        node = board.nodes.get(ann.nodo_id)
        if not node:
            continue
        nx, ny = node_xy(board, node.columna, node.row)
        x, y = annotation_xy(nx, ny, ann.slot)
        elements.append(
            base(
                f"an{ann.id}",
                "#fde68a",
                kind="note",
                x=x,
                y=y,
                text=ann.texto,
                tag="Nota",
                tone="amber",
            )
        )

    for arrow in board.arrows.values():
        source = board.nodes.get(arrow.de)
        target = board.nodes.get(arrow.a)
        if not source or not target:
            continue
        x1, y1, x2, y2 = _arrow_points(board, source, target)
        elements.append(
            base(arrow.id, INK, kind="arrow", x1=x1, y1=y1, x2=x2, y2=y2)
        )
        if arrow.label:
            elements.append(
                base(
                    f"al{arrow.id}",
                    INK,
                    kind="text",
                    x=(x1 + x2) / 2 - 40,
                    y=(y1 + y2) / 2 - 20,
                    text=arrow.label,
                )
            )

    return elements

File: web/src/lib/architectBoard.ts
import { Graph, layout } from "@dagrejs/dagre";

import type { BoardElement } from "@/components/canvas/boardElements";
import type { StickyTone } from "@/components/canvas/StickyNote";

/** Which kind of thing a node is — sets the sticky's colour and tag. */
export type NodeKind = "idea" | "decision" | "tarea" | "duda";

/** Every op the backend Architect can emit. Mirrors shared/protocol.ts. Every
 *  op carries the pixel coordinates the reducer needs; the model never sees
 *  pixels, the backend fills them in. */
export type ArchitectOp =
  | {
      type: "crear_nodo";
      id: string;
      texto: string;
      columna: number;
      kind: NodeKind;
      x: number;
      y: number;
    }
  | { type: "editar_nodo"; id: string; texto: string }
  | {
      type: "mover_nodo";
      id: string;
      columna: number;
      x: number;
      y: number;
    }
  | {
      type: "conectar";
      id: string;
      de: string;
      a: string;
      label?: string;
      x1: number;
      y1: number;
      x2: number;
      y2: number;
    }
  | {
      type: "pegar_nota";
      id: string;
      nodo_id: string;
      texto: string;
      autor?: string;
      x: number;
      y: number;
    }
  | { type: "borrar"; id: string }
  | {
      type: "titular_columna";
      columna: number;
      titulo: string;
      x: number;
      y: number;
    };

type NodeState = {
  texto: string;
  kind: NodeKind;
  columna: number;
  x: number;
  y: number;
};

type ArrowState = {
  de: string;
  a: string;
  label: string;
  x1: number;
  y1: number;
  x2: number;
  y2: number;
};

type TitleState = { titulo: string; x: number; y: number };

/** Annotations live in a side panel, not on the canvas, so the reducer for
 *  the pizarra does not carry them. `useSessionStream` peels them off the
 *  event stream on its own and pushes them into its own notes list. */
export type ArchitectState = {
  nodes: Map<string, NodeState>;
  arrows: Map<string, ArrowState>;
  titles: Map<number, TitleState>;
};

export function emptyState(): ArchitectState {
  return {
    nodes: new Map(),
    arrows: new Map(),
    titles: new Map(),
  };
}

const KINDS: NodeKind[] = ["idea", "decision", "tarea", "duda"];

// Kind → sticky style. Same palette the backend uses in architect_board.py, so
// a node from an op looks the same as one from the /digest bootstrap.
const STYLE: Record<
  NodeKind,
  { tag: string; tone: StickyTone; colour: string }
> = {
  idea: { tag: "Idea", tone: "amber", colour: "#f59e0b" },
  decision: { tag: "Decisión", tone: "green", colour: "#22c55e" },
  tarea: { tag: "Tarea", tone: "blue", colour: "#3b82f6" },
  duda: { tag: "Duda", tone: "violet", colour: "#8b5cf6" },
};

const INK = "#525252";

// Sticky size. Kept in sync with backend NOTE_W/NOTE_H so an op that carries
// pixel coords still lines up when dagre relaxes into similar positions.
const NOTE_W = 160;
const NOTE_H = 110;
// Where the flow starts on the canvas. Leaves the tool rail on the left with
// clear space, and pushes the diagram below the TopBar.
const ORIGIN_X = 210;
const ORIGIN_Y = 104;
// Room between ranks and between nodes in a rank. Wide enough that arrow
// labels can sit between two nodes without landing on one.
const RANK_SEP = 90;
const NODE_SEP = 46;

/** Runs dagre over the current architect graph and returns each node's
 *  top-left position, in board pixels. Left-to-right flow, so the room reads
 *  the pizarra the way a diagram reads: start on the left, arrows to the
 *  right. Called on every state change through deriveElements. */
function layoutFlow(state: ArchitectState): Map<string, { x: number; y: number }> {
  const positions = new Map<string, { x: number; y: number }>();
  if (state.nodes.size === 0) return positions;

  const g = new Graph({ multigraph: false, compound: false });
  g.setGraph({
    rankdir: "LR",
    ranksep: RANK_SEP,
    nodesep: NODE_SEP,
    marginx: 20,
    marginy: 20,
  });
  g.setDefaultEdgeLabel(() => ({}));

  for (const id of state.nodes.keys()) {
    g.setNode(id, { width: NOTE_W, height: NOTE_H });
  }
  for (const arrow of state.arrows.values()) {
    if (state.nodes.has(arrow.de) && state.nodes.has(arrow.a)) {
      g.setEdge(arrow.de, arrow.a);
    }
  }

  layout(g);

  for (const id of state.nodes.keys()) {
    const laid = g.node(id);
    if (!laid) continue;
    // Dagre gives centre coordinates; the renderer wants top-left.
    positions.set(id, {
      x: ORIGIN_X + laid.x - NOTE_W / 2,
      y: ORIGIN_Y + laid.y - NOTE_H / 2,
    });
  }
  return positions;
}

function copy(state: ArchitectState): ArchitectState {
  return {
    nodes: new Map(state.nodes),
    arrows: new Map(state.arrows),
    titles: new Map(state.titles),
  };
}

/** Applies one op and returns the new state. Ops on missing targets are
 *  ignored — the backend can emit them out of order across reconnects and the
 *  board would rather stay behind than throw. */
export function applyOp(state: ArchitectState, op: ArchitectOp): ArchitectState {
  const next = copy(state);
  switch (op.type) {
    case "crear_nodo": {
      if (!KINDS.includes(op.kind)) return state;
      next.nodes.set(op.id, {
        texto: op.texto,
        kind: op.kind,
        columna: op.columna,
        x: op.x,
        y: op.y,
      });
      return next;
    }
    case "editar_nodo": {
      const node = next.nodes.get(op.id);
      if (!node) return state;
      next.nodes.set(op.id, { ...node, texto: op.texto });
      return next;
    }
    case "mover_nodo": {
      const node = next.nodes.get(op.id);
      if (!node) return state;
      next.nodes.set(op.id, {
        ...node,
        columna: op.columna,
        x: op.x,
        y: op.y,
      });
      return next;
    }
    case "conectar": {
      next.arrows.set(op.id, {
        de: op.de,
        a: op.a,
        label: op.label ?? "",
        x1: op.x1,
        y1: op.y1,
        x2: op.x2,
        y2: op.y2,
      });
      return next;
    }
    case "pegar_nota": {
      // Notes go to a side panel, not the pizarra. The hook consumes this op
      // path separately; it never reaches the canvas reducer.
      return state;
    }
    case "borrar": {
      if (next.nodes.has(op.id)) {
        next.nodes.delete(op.id);
        // Arrows lose their anchor when a node goes. The backend cascades the
        // same way — this keeps the two ends aligned without extra ops.
        for (const [arid, arrow] of next.arrows) {
          if (arrow.de === op.id || arrow.a === op.id) next.arrows.delete(arid);
        }
        return next;
      }
      if (next.arrows.has(op.id)) {
        next.arrows.delete(op.id);
        return next;
      }
      return state;
    }
    case "titular_columna": {
      next.titles.set(op.columna, {
        titulo: op.titulo,
        x: op.x,
        y: op.y,
      });
      return next;
    }
    default:
      return state;
  }
}

/** The /digest bootstrap shape. Fields match architect_board.snapshot() on
 *  the backend. */
export type ArchitectSnapshot = {
  nodes: Array<{
    id: string;
    texto: string;
    columna: number;
    kind: NodeKind;
    x: number;
    y: number;
  }>;
  annotations: Array<{
    id: string;
    nodo_id: string;
    texto: string;
    autor?: string;
    x: number;
    y: number;
  }>;
  arrows: Array<{
    id: string;
    de: string;
    a: string;
    label?: string;
    x1: number;
    y1: number;
    x2: number;
    y2: number;
  }>;
  titles: Array<{ columna: number; titulo: string; x: number; y: number }>;
};

export function fromSnapshot(snapshot: ArchitectSnapshot): ArchitectState {
  const state = emptyState();
  for (const n of snapshot.nodes) {
    if (!KINDS.includes(n.kind)) continue;
    state.nodes.set(n.id, {
      texto: n.texto,
      kind: n.kind,
      columna: n.columna,
      x: n.x,
      y: n.y,
    });
  }
  // Annotations are read by useSessionStream directly, not stored here.
  for (const ar of snapshot.arrows) {
    state.arrows.set(ar.id, {
      de: ar.de,
      a: ar.a,
      label: ar.label ?? "",
      x1: ar.x1,
      y1: ar.y1,
      x2: ar.x2,
      y2: ar.y2,
    });
  }
  for (const t of snapshot.titles) {
    state.titles.set(t.columna, { titulo: t.titulo, x: t.x, y: t.y });
  }
  return state;
}

/** Turns the structured board into the flat elements the BoardLayer renders.
 *  Runs a left-to-right dagre layout on the graph so a change to any node or
 *  arrow reflows the diagram — the wire coords the backend sent are ignored
 *  here in favour of a layout that keeps the flow readable as it grows. */
export function deriveElements(state: ArchitectState): BoardElement[] {
  const els: BoardElement[] = [];
  const base = {
    width: 2,
    opacity: 100,
    dash: "solid" as const,
    radius: 8,
    cap: "round" as const,
  };

  const positions = layoutFlow(state);

  for (const [id, node] of state.nodes) {
    const style = STYLE[node.kind];
    const pos = positions.get(id) ?? { x: node.x, y: node.y };
    els.push({
      ...base,
      id: `n${id}`,
      color: style.colour,
      kind: "note",
      x: pos.x,
      y: pos.y,
      text: node.texto,
      tag: style.tag,
      tone: style.tone,
    });
  }

  for (const [id, arrow] of state.arrows) {
    const source = positions.get(arrow.de);
    const target = positions.get(arrow.a);
    if (!source || !target) continue;
    // Left-to-right routing: right edge of source, left edge of target. Any
    // rank difference dagre laid out is handled by the diagonal itself.
    const x1 = source.x + NOTE_W;
    const y1 = source.y + NOTE_H / 2;
    const x2 = target.x;
    const y2 = target.y + NOTE_H / 2;
    els.push({
      ...base,
      id,
      color: INK,
      kind: "arrow",
      x1,
      y1,
      x2,
      y2,
    });
    if (arrow.label) {
      els.push({
        ...base,
        id: `al${id}`,
        color: INK,
        kind: "text",
        x: (x1 + x2) / 2 - 40,
        y: (y1 + y2) / 2 - 20,
        text: arrow.label,
      });
    }
  }

  return els;
}

/** Guards the incoming op event so a malformed payload never reaches the
 *  reducer. Kind and column ranges match the backend. */
export function isArchitectOp(v: unknown): v is ArchitectOp {
  if (typeof v !== "object" || v === null) return false;
  const op = v as Record<string, unknown>;
  const t = op.type;
  if (typeof t !== "string") return false;
  switch (t) {
    case "crear_nodo":
      return (
        typeof op.id === "string" &&
        typeof op.texto === "string" &&
        typeof op.columna === "number" &&
        typeof op.kind === "string" &&
        KINDS.includes(op.kind as NodeKind) &&
        typeof op.x === "number" &&
        typeof op.y === "number"
      );
    case "editar_nodo":
      return typeof op.id === "string" && typeof op.texto === "string";
    case "mover_nodo":
      return (
        typeof op.id === "string" &&
        typeof op.columna === "number" &&
        typeof op.x === "number" &&
        typeof op.y === "number"
      );
    case "conectar":
      return (
        typeof op.id === "string" &&
        typeof op.de === "string" &&
        typeof op.a === "string" &&
        typeof op.x1 === "number" &&
        typeof op.y1 === "number" &&
        typeof op.x2 === "number" &&
        typeof op.y2 === "number"
      );
    case "pegar_nota":
      return (
        typeof op.id === "string" &&
        typeof op.nodo_id === "string" &&
        typeof op.texto === "string" &&
        typeof op.x === "number" &&
        typeof op.y === "number"
      );
    case "borrar":
      return typeof op.id === "string";
    case "titular_columna":
      return (
        typeof op.columna === "number" &&
        typeof op.titulo === "string" &&
        typeof op.x === "number" &&
        typeof op.y === "number"
      );
    default:
      return false;
  }
}
